feat(compute): add compute consume reservation samples

TetyanaYahodska opened the pull request:
## Description
Sample in NodeJs - https://github.com/GoogleCloudPlatform/nodejs-docs-samples/pull/3877/files

Fixes #<ISSUE-NUMBER>

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist

- [x] I have followed [Sample Format Guide](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md)
- [ ] `pom.xml` parent set to latest `shared-configuration`
- [ ] Appropriate changes to README are included in PR
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] **Tests** pass:   `mvn clean verify` **required**
- [x] **Lint**  passes: `mvn -P lint checkstyle:check` **required**
- [ ] **Static Analysis**:  `mvn -P lint clean compile pmd:cpd-check spotbugs:check` **advisory only**
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample 
- [ ] Please **merge** this PR for me once it is approved

minherz reviewed (changes requested):
please modify the code samples to maintain **a single** code sample per file/class.

TetyanaYahodska commented:
These samples demonstrate how to create instances with different reservation consumption options. So, changes are required for both the instance and reservation. 
Documentation for these three samples:
https://cloud.google.com/compute/docs/instances/reservations-consume#consuming_a_specific_shared_reservation, https://cloud.google.com/compute/docs/instances/reservations-consume#consuming_a_specific_single-project_reservation, https://cloud.google.com/compute/docs/instances/reservations-consume#consuming_instances_from_any_matching_reservation

minherz commented on compute/cloud-client/src/main/java/compute/reservation/ConsumeAnyMatchingReservation.java:
```diff
@@ -0,0 +1,184 @@
+package compute.reservation;
+
+// [START compute_consume_any_matching_reservation]
+
```
remove empty line

TetyanaYahodska replied:
Done

minherz reviewed (changes requested):
As I explained earlier we have a policy of having a single code sample per file. Having two methods violates this requirement. If you want to demonstrate a workflow that involves executing multiple methods and different client libraries you can do it in the single method that examples the workflow.
If you have further questions, please schedule a meeting and argument your case.

TetyanaYahodska commented:
In this PR, I fixed the empty lines before the tags for the samples created earlier, removed the cleanup methods from the setup methods, and fixed other issues based on the recommendations I received. The idea behind this was to keep our code consistent.

minherz commented on compute/cloud-client/src/test/java/compute/reservation/ConsumeReservationsIT.java:
```diff
@@ -0,0 +1,164 @@
+
+  private static final String PROJECT_ID = System.getenv("GOOGLE_CLOUD_PROJECT");
+  private static final String ZONE = "us-central1-a";
+  private static final String  RESERVATION_NAME = "test-reservaton-" + UUID.randomUUID();
+  private static final String INSTANCE_FOR_SPR = "test-instance-for-spr-" + UUID.randomUUID();
+  private static final String INSTANCE_FOR_ANY_MATCHING = "test-instance-" + UUID.randomUUID();
+  private static final String SPECIFIC_SHARED_INSTANCE =
+      "test-instance-shared-" + UUID.randomUUID();
```
nit: recommend to use the same random ID; it helps to troubleshoot failed tests

TetyanaYahodska replied:
Thank you! Fixed

minherz commented on compute/cloud-client/src/test/java/compute/reservation/CrudOperationsReservationIT.java:
```diff
@@ -111,7 +83,6 @@ public void testGetReservation()
 
   @Test
-  @Order(4)
   public void testListReservation() throws IOException {
```
assert on `RESERVATION_NAME` in line 91 (I cannot comment on untouched code lines). Asserting on prefix will pass the test in event there is reservation from a parallel tests running there.

TetyanaYahodska replied:
You are right, We can't predict reservations' order. That's why I'm testing prefix. Because the goal of the listReservations() method is to get list with reservations. Actually, we can delete this assertion and left only isNotNull() assertion. Would you recommend to delete this assertion?

minherz replied:
I suggest to use a known reservation name that is supposed to be there.
If we cannot do it then I prefer the validation that you suggest because both of them can be false positive but your suggestion let's us lose enforcing the order which makes testing run faster.

TetyanaYahodska replied:
As the purpose of this test is list the reservations and not give as some defined reservation, I prefer to left it as is.

minherz commented on compute/cloud-client/src/test/java/compute/reservation/ReservationIT.java:
nit: please review the use of the test order by this class.

minherz reviewed (commented):
Thorough work. Thank you. Everything looks good.  I've left a few of nit comments for your considerations.

Besides the comments, before merging these changes, please check the following two items:

1. These tests look heavy to me. They create and delete a lot of VM instances and templates. Please, consider opening a feature request and suggesting more lightweight method of testing these code samples. Keep it mind that we do not need to test that client libraries used in these samples work as expected. We only need to test that our code samples use these client libraries as expected.

1. It looks like three different code samples (ConsumeAnyMatchingReservation, ConsumeSingleProjectReservation and ConsumeSpecificSharedReservation) differ only by the value the object `ReservationAffinity` that is used when instance resource is created. Maintaining all of the is a clear overhead. Consider replacing these code samples with the code that showing how each of the different reservation affinity can be created.

minherz reviewed (approved):
Before merging, please have a look at my reply to [this comment thread](https://github.com/GoogleCloudPlatform/java-docs-samples/pull/9596#discussion_r1826306998). No actions required but consider making changes discussed in the thread.
